File: backend/src/main/java/com/gxs/bank/service/UpiService.java
package com.gxs.bank.service;

import com.gxs.bank.exception.BadRequestException;
import com.gxs.bank.exception.ResourceNotFoundException;
import com.gxs.bank.model.SavingsAccount;
import com.gxs.bank.model.UpiId;
import com.gxs.bank.model.User;
import com.gxs.bank.repository.AccountRepository;
import com.gxs.bank.repository.UpiIdRepository;
import com.gxs.bank.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UpiService {
    private final UpiIdRepository upiIdRepository;
    private final UserRepository userRepository;
    private final AccountRepository accountRepository;

    public UpiId registerUpi(UUID userId, UUID accountId, String requestedId) {
        User user = userRepository.findById(userId).orElseThrow();
        SavingsAccount account = accountRepository.findById(accountId).orElseThrow();
        if (!account.getUser().getId().equals(userId)) throw new BadRequestException("Unauthorized");
        
        UpiId upi = UpiId.builder()
            .user(user)
            .account(account)
            .upiAddress(requestedId + "@gxs")
            .isPrimary(true)
            .isActive(true)
            .build();
        return upiIdRepository.save(upi);
    }
    
    public List<UpiId> getUserUpis(UUID userId) {
        return upiIdRepository.findByUserId(userId);
    }
}

File: backend-python/com/gxs/bank/service/UpiService.py
from com.gxs.bank.exception.BadRequestException import BadRequestException
from com.gxs.bank.model.UpiId import UpiId
from com.gxs.bank.repository.AccountRepository import AccountRepository
from com.gxs.bank.repository.UpiIdRepository import UpiIdRepository
from com.gxs.bank.repository.UserRepository import UserRepository


class UpiService:
    def __init__(self, db):
        self.db = db
        self.upiIdRepository = UpiIdRepository(db)
        self.userRepository = UserRepository(db)
        self.accountRepository = AccountRepository(db)

    def registerUpi(self, userId: str, accountId: str, requestedId: str):
        user = self.userRepository.findById(userId)
        account = self.accountRepository.findById(accountId)
        if user is None or account is None:
            raise BadRequestException("Unauthorized")
        if account.userId != userId:
            raise BadRequestException("Unauthorized")

        upi = UpiId(
            user=user,
            userId=user.id,
            account=account,
            accountId=account.id,
            upiAddress=f"{requestedId}@gxs",
            isPrimary=True,
            isActive=True,
        )
        self.upiIdRepository.save(upi)
        self.db.commit()
        self.db.refresh(upi)
        return upi

    def getUserUpis(self, userId: str):
        return self.upiIdRepository.findByUserId(userId)
